src/library/compiler/ir_interpreter.cpp

```diff
@@ -0,0 +1,753 @@
+            return box(tag);
+        } else {
+            object *o = alloc_cnstr(tag, size, usize * sizeof(void *) + ssize);
+            for (size_t i = 0; i < args.size(); i++) {
+                cnstr_set(o, i, eval_arg(args[i]));
+            }
+            return o;
+        }
+    }
+
+    object * eval_expr(expr const & e, type t) {
+        switch (expr_tag(e)) {
+            case expr_kind::Ctor:
+                return alloc_ctor(expr_ctor_info(e), expr_ctor_args(e));
+            case expr_kind::Reset: { // release fields if unique reference in preparation for `Reuse` below
+                object * o = var(expr_reset_obj(e));
+                if (is_exclusive(o)) {
+                    for (size_t i = 0; i < expr_reset_num_objs(e).get_small_value(); i++) {
+                        cnstr_release(o, i);
+                    }
+                    return o;
+                } else {
+                    dec_ref(o);
+                    return box(0);
+                }
+            }
+            case expr_kind::Reuse: { // reuse dead allocation if possible
+                object * o = var(expr_reuse_obj(e));
+                // check if `Reset` above had a unique reference it consumed
+                if (is_scalar(o)) {
+                    // fall back to regular allocation
+                    return alloc_ctor(expr_reuse_ctor(e), expr_reuse_args(e));
+                } else {
+                    // create new constructor object in-place
+                    if (expr_reuse_update_header(e)) {
+                        cnstr_set_tag(o, ctor_info_tag(expr_reuse_ctor(e)).get_small_value());
+                    }
+                    for (size_t i = 0; i < expr_reuse_args(e).size(); i++) {
+                        cnstr_set(o, i, eval_arg(expr_reuse_args(e)[i]));
+                    }
+                    return o;
+                }
+            }
+            case expr_kind::Proj: // object field access
+                return cnstr_get(var(expr_proj_obj(e)), expr_proj_idx(e).get_small_value());
+            case expr_kind::UProj: // USize field access
+                return box_size_t(cnstr_get_usize(var(expr_uproj_obj(e)), expr_uproj_idx(e).get_small_value()));
+            case expr_kind::SProj: { // other unboxed field access
+                size_t offset = expr_sproj_idx(e).get_small_value() * sizeof(void *) +
+                                expr_sproj_offset(e).get_small_value();
+                object *o = var(expr_sproj_obj(e));
+                switch (t) {
+                    case type::Float: throw exception("floats are not supported yet");
+                    case type::UInt8: return box(cnstr_get_uint8(o, offset));
+                    case type::UInt16: return box(cnstr_get_uint16(o, offset));
+                    case type::UInt32: return box_uint32(cnstr_get_uint32(o, offset));
+                    case type::UInt64: return box_uint64(cnstr_get_uint64(o, offset));
```

Comment:
I may have have asked this before, but what was the reason for unboxed type never being boxed into tagged pointers? I.e. using `Nat`'s boxing for all integral types.

Reply:
I am assuming that if someone is using unboxed types, they really care about details, and want a precise cost model for how these values are stored in memory when boxed. It seems unnatural to store a big `UInt64` in a GMP `mpz` which contains a lot of unnecessary overhead. Most users don't even know how a GMP `mpz` value is represented.

Reply:
Okay, makes sense. We could of course use a different ABI to store these values in the interpreter (e.g. store them un-encoded in the `object *` on 64-bits), but that would again make the interpreter quite a bit more complex. It's probably safe to say that code using unboxed values should be compiled anway.

Reply:
> We could of course use a different ABI to store these values in the interpreter (e.g. store them un-encoded in the object * on 64-bits)

Not sure I understood your proposal. I want to point out that Interpreted code may interact with compiled code. That is, we must be able to create boxed scalar values in the interpreter and "send" them to compiled code.

Reply:
Yes, we would have to translate between the ABIs at interpreter/compiler boundaries. Which could end up being both more complex and costly if the boundary is crossed frequently.